feat(common): new pipe to support object for ngFor

harunurhan opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

Add keys_pipe which uses Object.keys in order to transform
an object to array of its keys.

Fix #2246

harunurhan commented:
I signed it!

vicb commented on modules/@angular/common/test/pipes/keys_pipe_spec.ts:
```diff
@@ -0,0 +1,25 @@
+    var keys: Array<string>;
+    var pipe: KeysPipe;
+
+    beforeEach(() => {
```
what about `null` & `undefined`

harunurhan replied:
Since there is no custom logic in pipe to handle `null` & `undefined`. I thought it wouldn't be necessary.
@vicb

In both cases, `Uncaught TypeError: Cannot convert undefined or null to object` is thrown by `Object.keys`

So do you still think that we need tests for them, or even do you think we should handle these in pipe and throw a different error with more meaningful message?

vicb replied:
yes

harunurhan replied:
I am sorry to ask again,

but was this "yes" meant to for only adding tests for `null` and `undefined` or also changing the pipe to handle these below, in addition to tests.
```
if (!value) {
  throw new InvalidPipeArgumentError(KeysPipe, value);
}
```

vicb commented on modules/@angular/common/src/pipes/keys_pipe.ts:
```diff
@@ -0,0 +1,19 @@
+ */
+@Pipe({name: 'keys', pure:false})
+export class KeysPipe implements PipeTransform {
+    transform(value: Object): Array<string> { return Object.keys(value); }
```
Do we really need a Pipe ? vs `let k of Object.keys(obj)`

harunurhan replied:
If you do `Object.keys(obj)` in template, you get `TypeError: Cannot read property 'keys' of undefined`. So it can be done by wrapping it up in a method of a component and calling this method from the template which is IMHO not a very good way. 

Also angular has `json` pipe which basically returns `JSON.stringify` as well. That's I assumed this is OK to have.


In addition if you look at #2246 and http://stackoverflow.com/questions/31490713/iterate-over-typescript-dictionary-in-angular-2 there is a demand for such a pipe.

vicb replied:
yep, you're absolutely right

vicb commented on modules/@angular/common/src/pipes/common_pipes.ts:
```diff
@@ -37,4 +38,5 @@ export const COMMON_PIPES = /*@ts2dart_const*/[
   ReplacePipe,
   I18nPluralPipe,
   I18nSelectPipe,
+  KeysPipe
```
add trailing comma

vicb commented:
please run `gulp format`

harunurhan commented:
Addressed all of your comment @vicb 

Just let me know if I should add 
```
if (!value) {
  throw new InvalidPipeArgumentError(KeysPipe, value);
}
```
to the beginning of `transform`, instead of letting `Object.keys` throw a `TypeError`.

IgorMinar commented:
We should discuss the performance impact of this pipe and if nothing else at least document that this is not the most efficient way of iterating over an object.

Fank commented:
@IgorMinar I prefer to add this pipe but only if the performance info were added you talked about.
Because in some cases you need this, i can list them if you want.

harunurhan commented:
I rebased and updated the code to be like the other pipes since they are changed recently.

But I am not sure what to add in docs for the performance warning.

gnujeremie commented on modules/@angular/examples/common/pipes/ts/keys_pipe.ts:
```diff
@@ -0,0 +1,23 @@
+@Component({
+  selector: 'json-pipe',
+  template: `<div *ngFor="let key of object | keys">
+    key is {{key}} and value is {{myObject[key]}}
```
Did I miss something, or shouldn't it be ```key is {{key}} and value is {{object[key]}}```?

harunurhan replied:
You're right. I forgot to update it, thanks!

harunurhan replied:
fixed it and some other typos

gnujeremie commented:
Any update about the status of this PR ?

And another question, with this pipe, how do you combine it with the async pipe ?

ie : 
```
<ul>
     <li *ngFor="let key of keys$ | async | keys">?</li>
</ul>
```
I'm not sure ```keys$[key]``` will work.

(I'm currently using a custom pipe using this method http://stackoverflow.com/a/37431657/1182487, which enables me to combine it with | async)

harunurhan commented:
@gnujeremie 
Unfortunately `keys$[key]` doesn't work, since it wasn't my usecase I didn't think it would be good to return key-value pairs instead of only keys. But if reviewers and others think that this pipe should support it, I can update!

mhevery commented:
@vicb  / @vsavkin Could we merge this? What are the arguments against it?

IgorMinar commented:
We should get out of the box perf right and iteration order needs to be
well specified and have good default.

How about we tackle this in A4? Can we add it to the roadmap?

On Tue, Nov 29, 2016, 12:47 AM Miško Hevery <notifications@github.com>
wrote:

> @vicb <https://github.com/vicb> / @vsavkin <https://github.com/vsavkin>
> Could we merge this? What are the arguments against it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/11319#issuecomment-263432007>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6J4_Od-EEgofxktALmVKlBJF6tPOks5rC2gkgaJpZM4J0gqc>
> .
>

mhevery commented:
@IgorMinar added to roadmap

ChrisProlls commented:
So any news about it ?

m98 commented:
When will this pipe add to the core? (I see it merged, but it's still not accessible) and it's not in [list of pipes](https://angular.io/api?type=pipe) on the website

harunurhan commented:
@m98 still not merged unfortunately.

I can rebase once again when it's ready to be merged.

mhevery commented on modules/@angular/common/test/pipes/keys_pipe_spec.ts:
```diff
@@ -0,0 +1,34 @@
+      });
+
+      it('should throw error if value is undefined',
+         () => { expect(() => pipe.transform(undefined)).toThrowError(); });
```
I think a better user ergonomics would be to return empty array.

mhevery commented on modules/@angular/common/test/pipes/keys_pipe_spec.ts:
```diff
@@ -0,0 +1,34 @@
+      it('should throw error if value is undefined',
+         () => { expect(() => pipe.transform(undefined)).toThrowError(); });
+
+      it('should throw error if value is null',
```
same here

harunurhan committed:
feat(common): new pipe to support object for ngFor

Add entries_pipe which uses Object.keys in order to transform
an object to array of its {key, value} pairs

Fix #2246

jotatoledo commented:
@gnujeremie made a valid observation. One of the most frecuently used pipes is async, and the current implementation of this new pipe doesnt work well with it. IMO would be better if the return of the type was Tuple<string,Object>[]; or something similar, allowing use cases like:

```
<ul>
     <li *ngFor="let tuple of keys$ | async | keys">
         <p>Key: {{ tuple.key }}</p>
         <p>Value: {{ tuple.value  }}</p>
     </li>
</ul>
```

harunurhan commented:
@jotatoledo I agree that it is a good idea to support both usecases, I was thinking `[string, any][]` and call the pipe `entries` but I would like to hear what core maintainers think.

@mhevery ?

mhevery commented:
I like the `{key: string, value: any}` approach.

harunurhan commented:
@mhevery Why not using `[string, any][]`, and later possibly adapt `Object.entries` without any API change and extra `map`  ?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

trotyl commented:
@harunurhan Since array destruction not supported in template yet, writing `tuple[0]` and `tuple[1]` would be awkward.

harunurhan commented:
@trotyl I wasn't aware of that. Yes writing `tuple[0]`, `tuple[1]` is indeed not ideal.

harunurhan commented:
@mhevery addressed comments

IgorMinar commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,33 @@
+      return [];
+    }
+
+    return Object.keys(object).map(key => ({key, value: object[key]}));
```
this will result in a new array being created every time change detection runs resulting in very bad performance and lots of memory pressure.

if we are going to do this, then we should make it more efficient by reusing the array and even individual key-value objects if the original object identify and it's keys remained unchanged.

take a look at ngRepeat from AngularJS - https://github.com/angular/angular.js/blob/c0adcc3a4b561260d0f1ff6b834ca940bf1971a9/src/ng/directive/ngRepeat.js#L533

some of the optimizations used there can be useful in understanding how to make this pipe more efficient.

in general what I'd like to see is that if the object identity hasn't changed we should return the same array. Also if the identity hasn't changed but only one or more keys were added/remove, we should return the same array containing the same objects for all the unchanged keys and only add/remove objects for changed keys.

IgorMinar commented on packages/common/src/pipes/entries_pipe.ts:
```diff
@@ -0,0 +1,33 @@
+      return [];
+    }
+
+    return Object.keys(object).map(key => ({key, value: object[key]}));
```
additionally, we can't leave the sort order to be unspecified. as soon as this gets released, there is going to be an issue filed, that will claim that the sort order is wrong for some usecase. I suggest that by default we sort the keys.

harunurhan replied:
well, order is specified (in a way), it's as same as `for...in` excluding keys in the prototype chain. How do you think they should be sorted?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys

IgorMinar replied:
from: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

>  There is no guarantee that for...in will return the indexes in any particular order. 

I'd do something like `Object.keys(object).sort().map` but then I'd rewrite all of that with the perf suggestions I mentioned above, which makes the implementation a bit tricky.

IgorMinar commented on packages/examples/common/pipes/ts/entries_pipe.ts:
```diff
@@ -0,0 +1,21 @@
+
+// #docregion EntriesPipe
+@Component({
+  selector: 'entries-pipe',
```
can you please rename this component to ExampleEntriesPipe and example-entries-pipe. otherwise the selector suggests that component is a pipe, which is not the case.

harunurhan replied:
I agree but I name it like that in favor of  consistency within the codebase, other pipe examples that I checked are also like that.

IgorMinar replied:
I see. Yeah. We need to rename the others as well. so feel free to keep it as is and we'll rename everything in one go later.

Toxicable commented:
@harunurhan Hows it going? Do you need a hand with the Performance aspect discussed above?

harunurhan commented:
@Toxicable that would be good, actually I tried a bit like 3 weeks, it was getting much larger than I expected to have and still had few problems. I didn't have time since then.

You can take it over if you can start before this weekend or I could have another look this weekend with your help.

Toxicable commented:
@harunurhan I wont have much free time till this weekend either.
Ping me on https://gitter.im/angular/angular with @Toxicable and we'll try sort it out then

Toxicable commented:
@harunurhan @IgorMinar 
I did a bit of tinkering this morning can came up with this. 
I've commented it with some thoughts/rational I had on perf as I was doing it.
Let me know what you think

https://stackblitz.com/edit/angular-gitter-rh6ofs?file=app/entries.pipe.ts

harunurhan commented:
@IgorMinar so where it got complex for me and what @Toxicable last solution doesn't have is keeping the same array reference with modifications after modifications to the same object without changing its identity. @Toxicable thinks that it's not what you meant in https://github.com/angular/angular/pull/11319#discussion_r169836170 . Can we make this clear?

Toxicable commented:
After talking with @alxhub We came up with the idea to use a input paramater of `pure`, while the pipe is still pure it can be toggled between pure and impure, with it being pure by default to encourage immutablity.
https://stackblitz.com/edit/entries-pipe?file=app%2Fentries.pipe.ts

mhevery reviewed (changes requested):
This is an EXCELENT start for #24200! Here are few things which have to happen before we can accept the PR.

- [ ] The Pipe needs to use [`KeyValueDiffers`](https://angular.io/api/core/KeyValueDiffers)
- [ ] The Pipe needs to not be pure so that it can store the return value from previous computation if the `KeyVaueDiffer` returns no changes. This will greatly reduce the memory pressure.
- [ ] The resulting array needs to be sorted in alphabetical order on `key`s.

If you can make the above changes, I will work with you to get this merged.

Toxicable commented:
@mhevery 
I was tinkering with this a little more and with the use of KeyValueDiffers I came up with 2 approaches.
Both can be seen here https://stackblitz.com/edit/entries-pipe-rndzk1?file=app%2Fentries.pipe.ts

1. If something has changed re generate the whole list.
- Always 3 iterations
- is quick and simple and will work every time
2. Try to diff the against the last value and only make new refs where needed
- Min 3 iterations, max 4
- is more complex but saves generating new pairs since it reuses more of the previous ones but it might possibly cause more iterations
-It's a memory vs CPU trade off depending on input changes.

(I may have incorrectly worked out those iterations so correct me if im wrong)

mhevery commented:
Let's do the simple implementation for now which will be smaller and easier to maintain, and if it becomes problem we can refactor.

Toxicable commented:
@mhevery I've not the author of this PR but i've opened up a new one here https://github.com/angular/angular/pull/24319 if that's preferable

Toxicable commented:
Can be closed via https://github.com/angular/angular/commit/2b49bf77afc22ae9b10323880d3e06efc783bc28

mhevery commented:
resolved by https://github.com/angular/angular/pull/24319
